File: solutions/890/890-kir3i.py
class Solution:
    def findAndReplacePattern(self, words, pattern):
        sz = len(pattern)
        ans = []
        for w in words:
            ok = True
            trans = {}
            used = set([])
            for i in range(sz):
                if pattern[i] not in trans:
                    if w[i] in used:
                        ok = False
                        break
                    trans[pattern[i]] = w[i]
                    used.add(w[i])
                elif trans[pattern[i]] != w[i]:
                    ok = False
                    break
            if ok:
               ans.append(w)
        return ans

File: solutions/890/890-kir3i.java
class Solution {
    public Map<Character, Character> trans = new HashMap<Character, Character>();
    public Set<Character> used = new HashSet<Character>();
    public List<String> findAndReplacePattern(String[] words, String pattern) {
        int sz = pattern.length();
        List<String> ans = new ArrayList<String>();
        for(String w: words) {
            Boolean ok = true;
            trans.clear();
            used.clear();
            for(int i=0; i<sz && ok; i++) {
                if(!trans.containsKey(pattern.charAt(i))) {
                    ok &= !used.contains(w.charAt(i));
                    trans.put(pattern.charAt(i), w.charAt(i));
                    used.add(w.charAt(i));
                }
                else {
                    ok &= (trans.get(pattern.charAt(i)) == w.charAt(i));
                }
            }
            if(ok)
                ans.add(w);
        }
        return ans;
    }
}
